### PR TITLE
Extend-product-service-with-type-fetching_CU-jqanj7

### PR DESCRIPTION
**Pull request by a-wilkes**

### CODE REVIEW

**Comment by a-wilkes** [comment]
Task linked: [CU-jqanj7 Extend product service with type fetching](https://app.clickup.com/t/jqanj7)

**Comment by lpboucher** [comment]
Task linked: [CU-jqanj7 Extend product service with type fetching](https://app.clickup.com/t/jqanj7)

**Comment by a-wilkes** [comment]
Since TS compiles down to JS and JS doesn't have types, I had to reintroduce `ProductType` which I don't feel is an ideal solution since it now means that there's a tight coupling between `ProductType` and the `Product` subclasses, as well as a redundancy in the information `ProductType` conveys. I'm not sure how else to do it, though, since you can't pass types as parameters. The only non-tightly coupled solution I could think of is a series of nearly identical getter methods like `getCoffees`, `getMerchandise` etc, which would only make sense if we can query the server-side API for these types (and even then, we'd couple the HTTP requests to the types). Am I over-thinking this, or am I missing a more elegant solution?

File: src/app/shared/models/classes/product.class.ts [hunk 1]
```diff
@@ -7,6 +8,7 @@ export abstract class Product {
     collection: CollectionType | null = null;
     description: string | null = null;
     slug: string | null = null;
+    abstract productType: ProductType;
```
Comment (lpboucher) [inline src/app/shared/models/classes/product.class.ts]: abstract properties should be at top of class

Comment (lpboucher) [reply src/app/shared/models/classes/product.class.ts]: done

File: src/app/core/products/product.service.ts [hunk 2]
```diff
@@ -118,4 +119,15 @@ export class ProductService {
                 take(2)
             );
     }
+
+    getProductsByType(productType?: ProductType | "all"): Observable<Product[]> {
+        if (productType === undefined || productType === "all") {
```
Comment (lpboucher) [inline src/app/core/products/product.service.ts]: undefined gets marshalled to false, so we can use the bang (!) operator

Comment (lpboucher) [reply src/app/core/products/product.service.ts]: done

**Review by lpboucher — APPROVED** [review]

**Comment by lpboucher** [comment]
Sorry, just saw this.

No, you have a good point! You are getting into more advanced Typescript features.

There are a few things we can do. Let’s discuss at a certain point.

You can check the data-api.service.ts and see how the return types include a dynamic parameter type (Observable<T>).

**Pull request merged**